src/backend/distributed/shared_library_init.c

```diff
@@ -388,19 +385,15 @@ _PG_init(void)
 	load_file(COLUMNAR_MODULE_NAME, false);
 
 	/*
-	 * Now, acquire symbols from columnar module. First, acquire
-	 * the address of the set options hook, and set it so that we
-	 * can propagate options changes.
+	 * Register utility hook. This must be done after loading columnar, so
+	 * that the citus hook is called first, followed by the columnar hook,
+	 * followed by standard_ProcessUtility. That allows citus to distribute
+	 * ALTER TABLE commands before columnar strips out the columnar-specific
+	 * options.
 	 */
-	ColumnarTableSetOptions_hook_type **ColumnarTableSetOptions_hook_ptr =
-		(ColumnarTableSetOptions_hook_type **) find_rendezvous_variable(
-			COLUMNAR_SETOPTIONS_HOOK_SYM);
-
-	/* rendezvous variable registered during columnar initialization */
-	Assert(ColumnarTableSetOptions_hook_ptr != NULL);
-	Assert(*ColumnarTableSetOptions_hook_ptr != NULL);
-
-	**ColumnarTableSetOptions_hook_ptr = ColumnarTableSetOptionsHook;
+	PrevProcessUtility = (ProcessUtility_hook != NULL) ?
+						 ProcessUtility_hook : standard_ProcessUtility;
```

Comment:
For Citus we introduce `PrevProcessUtility`, while for columnar we have `PrevProcessUtilityHook`. Both names are subtly different. I wonder if it would make sense to have specific `Columnar` and `Citus` in their names to avoid confusion.

I understand that both extensions are linked differently, and even the names are different. However it got me confused slightly while diving deep in the code that a) they were different, and b) which one is for Citus and Columnar. Mostly the file they are declared in dictates the difference, which was a bit hard, given they live in the same repo

Reply:
Adding "citus" or "columnar" to the name makes it more confusing to me, because it either implies that we know which extension the previous hook belongs to (which is a layering violation), or that it belongs to the current extension (which it doesn't).

Looking in `contrib/`, it appears that `prev_ProcessUtility` or `next_ProcessUtility` are the preferred names. I guess the "prev" and "next" are supposed to imply when we are calling it relative to our own code... though I had always read it as "preexisting_ProcessUtility". Honestly not quite sure what to do here and I don't want to over-think it right now.